frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseFastForwardParentToMatchBranchAction.java

```diff
@@ -0,0 +1,139 @@
+import com.virtuslab.gitmachete.backend.api.SyncToParentStatus;
+import com.virtuslab.gitmachete.frontend.actionids.ActionPlaces;
+import com.virtuslab.gitmachete.frontend.datakeys.DataKeys;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * Expects DataKeys:
+ * <ul>
+ *  <li>{@link DataKeys#KEY_GIT_MACHETE_REPOSITORY}</li>
+ *  <li>{@link CommonDataKeys#PROJECT}</li>
+ * </ul>
+ */
+public abstract class BaseFastForwardParentToMatchBranchAction extends GitMacheteRepositoryReadyAction
+    implements
+      IBranchNameProvider {
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  @Override
+  @UIEffect
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    var branchName = getNameOfBranchUnderAction(anActionEvent);
+    if (branchName.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Fast forward disabled due to undefined branch name");
+      return;
+    }
+
+    var gitMacheteBranch = getGitMacheteBranchByName(anActionEvent, branchName.get());
+    if (gitMacheteBranch.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Fast forward disabled due to undefined machete branch");
+      return;
+    }
+
+    if (gitMacheteBranch.get().isRootBranch()) {
+      if (anActionEvent.getPlace().equals(ActionPlaces.ACTION_PLACE_TOOLBAR)) {
+        presentation.setEnabled(false);
+        presentation.setDescription("Root branch '${branchName.get()}' cannot be fast-forwarded");
+      } else { //contextmenu
+        // in case of root branch we do not want to show this option at all
+        presentation.setEnabledAndVisible(false);
+      }
+      return;
+    }
+
+    var gitMacheteNonRoot = gitMacheteBranch.get().asNonRootBranch();
+    var syncToParentStatus = gitMacheteNonRoot.getSyncToParentStatus();
+
+    if (syncToParentStatus == SyncToParentStatus.InSync) {
+
+      if (getCurrentBranchNameIfManaged(anActionEvent).equals(branchName)) {
+        presentation.setText("Fast Forward Parent To Match Current Branch");
+      }
+
+      var parentName = gitMacheteNonRoot.getUpstreamBranch().getName();
+      presentation.setDescription("Fast forward branch '${parentName}' to match '${branchName.get()}'");
+
+    } else {
+      presentation.setEnabled(false);
+      var desc = Match(syncToParentStatus).of(
+          Case($(SyncToParentStatus.InSyncButForkPointOff), "in sync to its parent but fork point is off"),
+          Case($(SyncToParentStatus.MergedToParent), "merged into parent"),
+          Case($(SyncToParentStatus.OutOfSync), "out of sync to its parent"),
+          Case($(), "in unknown status '${syncToParentStatus.toString()}' to its parent"));
```

Comment:
IIRC this piece of code is also somewhere else. If it's true maybe extract it to utils?

Reply:
I would rather keep it like that. 
- this is the only place where it is used 
- we do not want to perform too much call within `update` (performance)